### PR TITLE
Transactional Batch Saving

### LINKED ISSUE #3694: saving skeletons: ensure that update action batches are saved transactionally
add transactionId, transactionLength, transactionIndex to update action groups,
on server, write them to temporary storage (possibly a redis db), when loading, show nothing of that
commit only when receiving the last one
reject if temporary storage contains previous one that is committed and has a different transactionId
whoever commits first wins

### PR DESCRIPTION
**Pull request by fm3**
### URL of deployed dev instance (used for testing):
- https://savebatchtransactions.webknossos.xyz

### Steps to test:
- drag large NML into wK
- save, leave page before it reaches 100%
- reload page, should get old or new tracing (not something in between)
- kill backend during saving
- retry should either continue seamlessly or show conflict and the upcoming reload should give the old (consistent) tracing – (which of these two depends on the timing of the backend kill)
- the only way to still get inconsistent tracings, i think, is to kill the backend during the committing phase when it received the final request of a transaction
- when having a tracing open in two tabs, one (A) saving a large NML, and one (B) saving a small change after A has already successfully sent some requests, the small change (B) should still win (commits first) and the batched large saving process from A should fail when committing (refresh should then yield the small change from B)
- normal saving should also still work (both skeleton + volume)
- normal saving combined with a large transaction should also work

### Issues:
- fixes #3694 

### TODO
- [x] sync batch cache to disk (redis?)
- [x] ignore duplicate requests properly again (see #3767 )
- [x] clean up code, extract functions?
- [x] remove redundant request id
- [x] add config for redis address
- [x] in health check, try redis
- [x] redis error handling?
- [x] update readme (redis is now required)
- [x] discuss if progress percentage should be adapted (committing requests are significantly slower on the server) (→for now, let’s leave it as is)
- [x] fix front-end tests
- [x] include redis in docker files
- [x] adapt infrastructure setup to include redis
- [x] test health on a dev instance
- [x] test all the scenarios
- [x] after testing, remove (some of) the debug output

------
- [x] Updated [changelog](../blob/master/CHANGELOG.md#unreleased)
- [x] Updated [migration guide](../blob/master/MIGRATIONS.md#unreleased) if applicable
- ~~[ ] Updated [documentation](../blob/master/docs) if applicable~~
- [x] Needs datastore update after deployment
- [x] Ready for review (backend)

### CODE REVIEW

**Comment by fm3** [comment]
@jstriebel could I ask you to look into adding redis to the docker files so that we can eventually use this in production? :) I’d also be grateful for a first review/look at the backend changes. I realize the code is rather complex and I’d be happy to discuss it further

@philippotto Could you have a look at the frontend tests? :)

File: webknossos-datastore/app/com/scalableminds/webknossos/datastore/storage/TemporaryStore.scala [hunk 1]
```diff
@@ -44,6 +44,13 @@ class TemporaryStore[K, V] @Inject()(system: ActorSystem) {
       }
     }
 
+  def findAllConditional(predicate: K => Boolean) =
+    map.synchronized {
+      map.keySet.filter(predicate).flatMap { key: K =>
+        map.get(key)
+      }
+    }
+
```
Comment (fm3) [inline webknossos-datastore/app/com/scalableminds/webknossos/datastore/storage/TemporaryStore.scala]: note to self: no longer used, remove

File: webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/controllers/TracingController.scala [hunk 2]
```diff
@@ -109,38 +110,93 @@ trait TracingController[T <: GeneratedMessage with Message[T], Ts <: GeneratedMe
...
-                    }
-                  }
-                }
+          if (updateGroups.forall(_.transactionGroupCount.getOrElse(1) == 1)) {
+            logger.debug("only single-group transactions in this request. committing all.")
+            commitUpdates(tracingId, updateGroups, userToken).map(_ => Ok)
+          } else {
+            updateGroups
+              .foldLeft(tracingService.currentVersion(tracingId)) { (currentCommittedVersionFox, updateGroup) =>
+                handleUpdategroupForTransaction(tracingId, currentCommittedVersionFox, updateGroup, userToken)
```
Comment (fm3) [inline webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/controllers/TracingController.scala]: updateGroup with capital G

File: README.md [hunk 3]
```diff
@@ -98,7 +99,7 @@ echo "deb https://dl.yarnpkg.com/debian/ stable main" | sudo tee /etc/apt/source
 
 # Installing everything
 sudo apt-get update
-sudo apt-get install -y git postgresql-10 postgresql-client-10 nodejs scala sbt openjdk-8-jdk yarn
+sudo apt-get install -y git postgresql-10 postgresql-client-10 nodejs scala sbt openjdk-8-jdk yarn redis
```
Comment (jstriebel) [inline README.md]: I think `redis-server` is currently the way to go (`redis` is only available since buster, for stretch only via backports)

Comment (jstriebel) [reply README.md]:
```suggestion
sudo apt-get install -y git postgresql-10 postgresql-client-10 nodejs scala sbt openjdk-8-jdk yarn redis-server
```

File: docker-compose.yml [hunk 4]
```diff
@@ -261,3 +261,11 @@ services:
     volumes:
       - "./fossildb-dev/data:/fossildb/data"
       - "./fossildb-dev/backup:/fossildb/backup"
+
+  # Redis
+  redis:
+    image: redis:5.0.3
```
Comment (jstriebel) [inline docker-compose.yml]: Let's use the same as in k8:
```suggestion
    image: redis:5.0-alpine
```

File: docker-compose.yml [hunk 5]
```diff
@@ -261,3 +261,11 @@ services:
...
+    volumes:
+      - "./redis-data:/data"
```
Comment (jstriebel) [inline docker-compose.yml]: Not sure if we need this to be persisted on-disk for local development / CI, I would rather avoid it. If it should still stay in, let's use this command:
```
- "redis-server"
- "--appendonly"
- "yes"
```

File: docker-compose.yml [hunk 6]
```diff
@@ -68,6 +69,7 @@ services:
       - -Ddatastore.oxalis.uri=webknossos:9000
     links:
       - fossildb-persisted:fossildb
+      - redis:redis
```
Comment (jstriebel) [inline docker-compose.yml]:
```suggestion
      - redis
```
This syntax is enough here, as the service-name and the dns entry match.

File: docker-compose.yml [hunk 7]
```diff
@@ -11,6 +11,7 @@ services:
     links:
       - "fossildb-persisted:fossildb"
       - "postgres-persisted:postgres"
+      - "redis:redis"
```
Comment (jstriebel) [inline docker-compose.yml]:
```suggestion
      - redis
```
This syntax is enough here, as the service-name and the dns entry match.

**Comment by fm3** [comment]
@jstriebel I think all checkboxes are ticked now, would you like to go for a final review round? :)

File: README.md [hunk 8]
```diff
@@ -119,6 +120,9 @@ See: http://www.scala-sbt.org/release/docs/Getting-Started/Setup.html
 * Install PostgreSQL from https://www.postgresql.org/download/
 * PostgreSQL version **10+ is required**
 
+##### PostgreSQL
```
Comment (jstriebel) [inline README.md]:
```suggestion
##### Redis
```
:trollface:

File: conf/application.conf [hunk 9]
```diff
@@ -61,7 +65,7 @@ braintracing {
 application {
   insertInitialData = true
   authentication {
-    enableDevAutoLogin = false
+    enableDevAutoLogin = true
```
Comment (jstriebel) [inline conf/application.conf]: Why is this changed in this PR?

File: webknossos-tracingstore/deployment/config/tracingstore-docker.conf [hunk 10]
```diff
@@ -23,4 +23,6 @@ tracingstore {
     address = "fossildb"
     port = 7155
   }
+  
+  redis.address = "redis"
```
Comment (jstriebel) [inline webknossos-tracingstore/deployment/config/tracingstore-docker.conf]: I just updated this in the deployment blueprint config.

File: conf/application.conf (hunk 9 above)
Comment (fm3) [reply conf/application.conf]: @daniel-wer I noticed you changed this to false in the application.conf. was that intentional?

Comment (daniel-wer) [reply conf/application.conf]: No that was not intentional, but I often had to disable it to test logged-out-behavior before, so I actually preferred it being set to `false`. Logging in is easy and quick. What's your opinion @philippotto?

Comment (philippotto) [reply conf/application.conf]: I agree with @daniel-wer and would prefer it being false, too.

Comment (fm3) [reply conf/application.conf]: fine by me :)

Comment (jstriebel) [reply conf/application.conf]: :+1:

File: frontend/javascripts/oxalis/model/sagas/save_saga.js [hunk 11]
```diff
@@ -186,6 +182,7 @@ export function* sendRequestToServer(tracingType: "skeleton" | "volume"): Saga<v
           method: "POST",
           headers: { "X-Date": `${Date.now()}` },
           data: compactedSaveQueue,
+          // todo: undo this before merging
```
Comment (daniel-wer) [inline frontend/javascripts/oxalis/model/sagas/save_saga.js]: Reminder to undo this before merging :)
